### PR TITLE
Fix final wasm backend pthread tests

### PR DESCRIPTION
**Pull request by kripken**
One issue was https://github.com/emscripten-core/emscripten/issues/8916 - `static EMSCRIPTEN_KEEPALIVE` works differently in upstream vs fastcomp. It should possibly be an error anyhow; I just fixed this in the test suite code (i.e. I removed `static`).

The other was https://github.com/emscripten-core/emscripten/issues/8917 - the constructor attribute is used in some pthreads library code to initialize something, and it's not run automatically. Adding it to the exported functions, and running it manually, fixes things. (If we think this should work, maybe that is just a temporary workaround.)

This PR fixes the last part of #8718 - after this, all pthreads tests should be enabled, except for those using `MAIN_THREAD_EM_ASM()` which we have not yet implemented there. But there haven't been requests for that, so I'm not sure how important it is.

### CODE REVIEW

**Comment by kripken** [comment]
The `report_result.h` file used `static` which I removed as mentioned earlier, and that uncovered a bad hack in the test suite - we use `-include` on it, which means it is added to each source file we build. If there is more than one, then without static we get duplicate symbols. To fix this, I split out the body into a cpp file.

File: tests/report_result.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2019 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ *
+ * Define REPORT_RESULT and REPORT_RESULT_SYNC for using in test code
```
Comment (sbc100) [inline tests/report_result.cpp]: This documentation seems wrong.

File: tests/runner.py [hunk 2]
```diff
@@ -1399,7 +1399,7 @@ def run_browser(self, html_file, message, expectedResult=None, timeout=None):
       print('(moving on..)')
 
   def with_report_result(self, code):
-    return '#define EMTEST_PORT_NUMBER %d\n#include "%s"\n' % (self.port, path_from_root('tests', 'report_result.h')) + code
+    return '#define EMTEST_PORT_NUMBER %d\n#include "%s"\n#include "%s"\n' % (self.port, path_from_root('tests', 'report_result.h'), path_from_root('tests', 'report_result.cpp')) + code
```
Comment (sbc100) [inline tests/runner.py]: Including the .cpp file like this also seems strange.

File: tests/runner.py [hunk 3]
```diff
@@ -1540,7 +1540,7 @@ def btest(self, filename, expected=None, reference=None, force_c=False,
     if 'WASM=0' not in args:
       # Filter out separate-asm, which is implied by wasm
       args = [a for a in args if a != '--separate-asm']
-    args += ['-DEMTEST_PORT_NUMBER=%d' % self.port, '-include', path_from_root('tests', 'report_result.h')]
+    args += ['-DEMTEST_PORT_NUMBER=%d' % self.port, '-include', path_from_root('tests', 'report_result.h'), path_from_root('tests', 'report_result.cpp')]
```
Comment (sbc100) [inline tests/runner.py]: Adding a .cpp file as a compiler flag seems a little off.

**Comment by sbc100** [comment]
Lets split out the report result change perhaps?

**Comment by kripken** [comment]
Good idea, I split off the locale stuff into #8945

File: tests/runner.py (hunk 2 above)
Comment (kripken) [reply tests/runner.py]: I agree it does look somewhat strange. I refactored this now to clarify the various pieces, and to embed the cpp there, let me know if that's clearer.

File: tests/runner.py (hunk 3 above)
Comment (kripken) [reply tests/runner.py]: I added a comment and cleaned this up a little - what do you think now?

I think it's ok here to add another source file to be compiled as another argument, with a comment explaining why, but maybe I don't understand what you don't like about it? It basically ends up with `emcc [original source files and arguments] another_source.cpp`, so it adds it to the existing sources.

File: tests/report_result.cpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    maybeReportResultToServer($0, $1, $2);
+  }, result, sync, EMTEST_PORT_NUMBER);
+}
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif // __EMSCRIPTEN__
+
```
Comment (sbc100) [inline tests/report_result.cpp]: Remove trailing newline

**Pull request merged**